File: Shaders/ShaderManager.ts
import GameManager from "GameManager";
import Shader from "./Shader";
import Asset from "Assets/Asset";

export default class ShaderManager {
  gameManager: GameManager;
  shaders: Map<string, Shader> = new Map<string, Shader>();

  constructor(gameManager: GameManager) {
    this.gameManager = gameManager;
  }

  addShader(assetId: string, shader: Shader): void {
    this.shaders.set(assetId, shader);
  }

  getShader(assetId: string): Shader | undefined {
    return this.shaders.get(assetId);
  }

  updateShaders(dt: number): void {
    for (const [id, shader] of this.shaders) {
      shader.updateShader(dt);
      if (shader.isFinished) {
        this.shaders.delete(id);
      }
    }
  }

  applyShaders(asset: Asset): void {
    const shader = this.shaders.get(asset.id);
    if (shader && shader.isShading) {
      love.graphics.setShader(shader.shader);
    }
  }

  removeShaders(): void {
    love.graphics.setShader();
  }
}

File: Shaders/ShaderManager.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__Iterator = ____lualib.__TS__Iterator
local ____exports = {}
____exports.default = __TS__Class()
local ShaderManager = ____exports.default
ShaderManager.name = "ShaderManager"
function ShaderManager.prototype.____constructor(self, gameManager)
    self.shaders = __TS__New(Map)
    self.gameManager = gameManager
end
function ShaderManager.prototype.addShader(self, assetId, shader)
    self.shaders:set(assetId, shader)
end
function ShaderManager.prototype.getShader(self, assetId)
    return self.shaders:get(assetId)
end
function ShaderManager.prototype.updateShaders(self, dt)
    for ____, ____value in __TS__Iterator(self.shaders) do
        local id = ____value[1]
        local shader = ____value[2]
        shader:updateShader(dt)
        if shader.isFinished then
            self.shaders:delete(id)
        end
    end
end
function ShaderManager.prototype.applyShaders(self, asset)
    local shader = self.shaders:get(asset.id)
    if shader and shader.isShading then
        love.graphics.setShader(shader.shader)
    end
end
function ShaderManager.prototype.removeShaders(self)
    love.graphics.setShader()
end
return ____exports
